### PR TITLE
[Feature][Op][MRV2] Optimize gumbel_sample with hierarchical categorical sampling on NPU

### PR DESCRIPTION
**Pull request by AuroraEmiya**
### What this PR does / why we need it?

This PR optimizes `gumbel_sample` for MRV2 on Ascend NPU by replacing the original vocabulary-wide Gumbel-Max path with a distribution-equivalent Triton `categorical_sample` implementation.

The original Gumbel-Max implementation generates one random value for every vocabulary element, applies the Gumbel transform, and then performs `argmax`:

```text
logits
  -> optional temperature scaling
  -> U_i ~ Uniform(0, 1) for every vocab element
  -> g_i = -log(-log(U_i))
  -> logits + g_i
  -> argmax
```

For large vocabularies such as `V=151936`, this is expensive on NPU because the random-number generation and transcendental operations scale with `B × V`.

For non-zero temperature, Gumbel-Max is distribution-equivalent to directly sampling from the categorical distribution defined by `softmax(logits / T)`. This PR therefore changes the implementation strategy while preserving the sampling probability distribution.

The optimized path uses a two-stage hierarchical categorical sampler:

```text
8192-element coarse block
        ↓
prepare block max / argmax / probability mass
        ↓
prepare 8 × 1024 fine-block masses
        ↓
sample one coarse block
        ↓
sample one 1024-element fine block
        ↓
reload only the selected 1024 logits
        ↓
final cumulative-mass search
        ↓
sampled token
```

The main implementation details are:

- Split each vocabulary row into **8192-element coarse blocks**.
- `_categorical_prepare_mass_kernel` computes:
  - block max / argmax,
  - coarse-block probability mass,
  - **8 × 1024 fine-block masses** for each coarse block.
- `_categorical_sample_kernel`:
  - samples one coarse block,
  - samples one prepared 1024-element fine block,
  - reloads only the selected 1024 logits for the final cumulative-mass search.
- Only **one stateless uniform random draw per sampled row** is required instead of one random value per vocabulary element.
- `(token, coarse_block)` prepare tasks and token sampling tasks are distributed according to the available Vector Core count.
- Draft sampling uses a separate stateless RNG namespace by adding `1 << 30` to the random position.
- Pre-imported sampler/speculator modules are explicitly rebound to `categorical_sample`, because those modules may be imported before the Ascend patch is applied.

The existing sampling contract is preserved:

- `temperature == 0`: greedy `argmax`.
- `temperature != 0`: categorical sampling from the same probability distribution as Gumbel-Max.
- `apply_temperature` behavior is preserved.
- `expanded_idx_mapping`, `seed`, `pos`, `logits_cache`, and `logits_cache_col` behavior is preserved.
- Raw logits are written to `logits_cache` before temperature scaling.
- CUDAGraph padding rows with request index `-1` do not write the cache.
- `use_fp64=True` remains unsupported on NPU and raises `NotImplementedError`.

This work follows the direct-categorical-sampling direction explored in #16176, while using an NPU-specific hierarchical 8K/1K implementation to reduce the final scan size and avoid vocabulary-wide random-number generation.

---

### Performance

Performance was measured with `msprof op`. The categorical operator is profiled as two Triton kernels:

```text
categorical total
=
_categorical_prepare_mass_kernel
+
_categorical_sample_kernel
```

The numbers below are kernel `Task Duration(us)` and do not include Python wrapper or synchronized host-side wall time.

#### Categorical scaling on A3, FP32, `temperature=1`

Batch growth with `V=151936`:

| Batch | Vocab | Prepare Mass | Sample | Total |
| ---: | ---: | ---: | ---: | ---: |
| 1 | 151936 | 7.08 us | 25.78 us | **32.86 us** |
| 16 | 151936 | 34.34 us | 24.20 us | **58.54 us** |
| 32 | 151936 | 62.40 us | 25.28 us | **87.68 us** |
| 64 | 151936 | 117.48 us | 46.56 us | **164.04 us** |
| 128 | 151936 | 227.44 us | 88.10 us | **315.54 us** |
| 256 | 151936 | 448.78 us | 150.22 us | **599.00 us** |
| 512 | 151936 | 889.58 us | 273.84 us | **1163.42 us** |

Vocabulary scaling with `B=64`:

| Batch | Vocab | Prepare Mass | Sample | Total |
| ---: | ---: | ---: | ---: | ---: |
| 64 | 32768 | 30.42 us | 45.76 us | **76.18 us** |
| 64 | 65536 | 51.84 us | 45.88 us | **97.72 us** |
| 64 | 131072 | 99.02 us | 45.58 us | **144.60 us** |
| 64 | 151936 | 117.48 us | 46.56 us | **164.04 us** |

The final sampling stage is largely decoupled from vocabulary size because only the selected 1024-element fine block is reloaded and scanned.

#### Previous NPU Gumbel-Max comparison

| Batch | Vocab | Previous Gumbel-Max | Categorical | Speedup |
| ---: | ---: | ---: | ---: | ---: |
| 1 | 151936 | 45366.88 us | 32.86 us | **~1380.6×** |
| 64 | 151936 | 90728.34 us | 164.04 us | **~553.1×** |

The main gain comes from eliminating vocabulary-wide random-number generation and the two logarithm operations required by the Gumbel transform.

A representative model workload also showed the sampling cost dropping from approximately `80324.612 us` to `155.53 us` on average, reducing its time ratio from `43.689%` to `0.13%` in that workload.

---

### Comparison with V1 `random_sample`

V1 `random_sample` uses exponential-race sampling:

```python
q = empty_exponential_noise_like(probs, use_fp64_gumbel)
q.exponential_()
scores = probs.div_(q)
return scores.argmax(dim=-1)
```

This is also distribution-equivalent to categorical sampling.

For the V1 comparison, `probs` is precomputed so that only the V1 `random_sample` core is profiled. Therefore the V1 numbers below **do not include softmax or other logits-to-probability preprocessing**, which intentionally makes the V1 baseline more favorable.

On Ascend A2/A3, `Tensor.exponential_()` is lowered into a sequence similar to:

```text
Uniform RNG
  -> Neg
  -> Add
  -> GreaterEqual
  -> MaskedFill
  -> Log
  -> Mul
  -> RealDiv
  -> ArgMax
  -> optional Cast
```

The current targeted `msprof op` run captured the exponential transform, division, `argmax`, and cast stages, but did not expose the underlying Uniform RNG kernel through the selected kernel-name path. Therefore the V1 values below are reported as **conservative captured lower bounds**: the actual V1 `random_sample` device cost is no lower than these values.

#### V1 core — batch growth at `V=151936`

| Batch | Vocab | V1 captured lower bound |
| ---: | ---: | ---: |
| 1 | 151936 | **58.80 us** |
| 16 | 151936 | **117.82 us** |
| 32 | 151936 | **176.14 us** |
| 64 | 151936 | **315.52 us** |
| 128 | 151936 | **660.70 us** |
| 256 | 151936 | **1635.60 us** |
| 512 | 151936 | **3755.52 us** |

Additional reviewer-oriented V1 anchors:

| Batch | Vocab | V1 captured lower bound |
| ---: | ---: | ---: |
| 16 | 32000 | **48.74 us** |
| 64 | 32000 | **95.32 us** |

#### V1 core — vocabulary growth at `B=64`

| Batch | Vocab | V1 captured lower bound |
| ---: | ---: | ---: |
| 64 | 32768 | **94.42 us** |
| 64 | 65536 | **157.58 us** |
| 64 | 131072 | **277.28 us** |
| 64 | 151936 | **315.52 us** |

Using the conservative V1 lower bound above, the long-vocabulary categorical results are already lower:

| Batch | Vocab | V1 captured lower bound | Categorical | Categorical speedup vs V1 lower bound |
| ---: | ---: | ---: | ---: | ---: |
| 1 | 151936 | 58.80 us | 32.86 us | **1.79×** |
| 16 | 151936 | 117.82 us | 58.54 us | **2.01×** |
| 32 | 151936 | 176.14 us | 87.68 us | **2.01×** |
| 64 | 151936 | 315.52 us | 164.04 us | **1.92×** |
| 128 | 151936 | 660.70 us | 315.54 us | **2.09×** |
| 256 | 151936 | 1635.60 us | 599.00 us | **2.73×** |
| 512 | 151936 | 3755.52 us | 1163.42 us | **3.23×** |

At `B=64`, the same trend is visible as vocabulary size grows:

| Vocab | V1 captured lower bound | Categorical | Categorical speedup vs V1 lower bound |
| ---: | ---: | ---: | ---: |
| 32768 | 94.42 us | 76.18 us | **1.24×** |
| 65536 | 157.58 us | 97.72 us | **1.61×** |
| 131072 | 277.28 us | 144.60 us | **1.92×** |
| 151936 | 315.52 us | 164.04 us | **1.92×** |

The gap grows with larger `B × V` because V1 exponential-race sampling still performs elementwise work over the full probability tensor, while the categorical implementation generates only one random value per row and restricts the final token search to one 1024-element fine block.

---

### Accuracy / probability-distribution validation

The optimized implementation changes the random-number consumption pattern, so exact sampled token sequences are not expected to match the previous Gumbel-Max implementation for the same `seed` and `pos`.

The correctness requirement is instead:

> the empirical sampling distribution must match the categorical distribution defined by the input logits and temperature.

The operator test includes a long-vocabulary business-shape distribution check:

```text
B = 64
V = 151936
trials = 256
total samples = 16384
```

The finite-probability support spans coarse-block, fine-block, and tail boundaries:

```text
[7, 1023, 1024, 8191, 8192, 65535, 131071, V - 1]
```

with support logits:

```text
[-1.5, -1.0, -0.5, 0.0, 0.5, 1.0, 1.5, 2.0]
```

The empirical probabilities are compared against `torch.softmax` and must satisfy:

```text
max absolute probability error <= 0.02
total variation distance       <= 0.04
```

This directly exercises:

- the `1024` fine-block boundary,
- the `8192` coarse-block boundary,
- long-vocabulary indexing,
- vocabulary tail handling,
- hierarchical probability-mass reconstruction.

An additional equal-mass random-sampling sanity case verifies that samples stay within the expected support and that all supported tokens are reached.

---

### Does this PR introduce _any_ user-facing change?

Yes, in sampling sequence behavior.

There is no new public sampling API, and the target probability distribution is unchanged. However, the internal random-sampling algorithm changes from Gumbel-Max to direct categorical sampling.

Because the two algorithms consume random numbers differently, the exact sampled token sequence for the same `seed` and `pos` is **not guaranteed to match** the previous implementation.

The following properties are preserved:

- greedy results for `temperature == 0`,
- categorical probability distribution for random sampling,
- deterministic replay within the new implementation for identical inputs,
- request-level temperature / seed behavior,
- raw-logits cache semantics,
- draft-vs-target RNG separation.

---

### How was this patch tested?

Added/updated operator and patch tests:

```text
tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_categorical_sample.py
tests/ut/patch/worker/test_patch_v2_gumbel.py
```

The categorical operator tests cover:

- FP32 / FP16 / BF16 greedy sampling.
- Exact greedy `argmax` behavior for `temperature=0`.
- Mixed greedy and random rows.
- `1024` fine-block boundaries.
- `8192` coarse-block boundaries.
- Non-aligned vocabulary tail at `V=151936`.
- Deterministic replay with identical `seed` and `pos`.
- Target and drafting RNG paths.
- `apply_temperature=True/False` equivalence against pre-scaled logits.
- Raw `logits_cache` semantics.
- Scalar and per-token `logits_cache_col`.
- CUDAGraph padding request mapping (`-1`) without invalid cache writes.
- Shared request mapping.
- Cache-write invariance of sampled output.
- Equal-mass random-distribution sanity.
- `B64 / V151936` empirical probability-distribution validation against `torch.softmax`.
- Unsupported `use_fp64=True` behavior.

The patch test additionally checks that modules imported before the Ascend patch are rebound to `categorical_sample`, including sampler/speculator consumers.

Example operator-test command:

```bash
pytest -sv tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_categorical_sample.py
```

Performance validation uses explicit target-kernel `msprof op` profiling rather than Python synchronized wall time.

---

### References

- Direct categorical sampling draft / related work: #16176
- AscendC categorical implementation referenced by that work: #14141

- RFC :https://github.com/vllm-project/vllm-ascend/issues/14130

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Review by Liamup777 — COMMENTED** [review]
<html><head></head><body><h3>Performance</h3><p>Tested <code inline="">_gumbel_sample</code> with the real workload in the end-to-end model.</p>

  | Count | Total Time (us) | Avg Time (us) | Min Time (us) | Max Time (us) | Ratio
-- | -- | -- | -- | -- | -- | --
Before | 54 | 4337529.071 | 80324.612  | 43802.251| 87636.067 | 43.689%
After | 54 | 8398.62| 155.53 | 32.320 | 162.671 | 0.13%

<p>The average kernel latency is reduced from <strong>80324.612 us to 155.53 us</strong>, achieving about <strong>518x speedup</strong>.</p><p>The kernel's end-to-end execution time ratio is reduced from <strong>43.689% to 0.13%</strong>.</p></body></html>

**Review by Ronald1995 — COMMENTED** [review]
please show your test result of precision for specific dataset.

**Review by Liamup777 — COMMENTED** [review]
precision test result: 
| dataset | version | metric | mode | pr-gumbel-s42-smoke | model runner version|
|----- | ----- | ----- | ----- | ----- | -----| 
| gsm8k | 50861a | accuracy | gen | 95.75 | V1 |
| gsm8k | 50861a | accuracy | gen | 95.6 | V2 |

**Comment by AuroraEmiya** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by AuroraEmiya** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by AuroraEmiya** [comment]
> precision test result:
> 
> dataset	version	metric	mode	pr-gumbel-s42-smoke	model runner version
> gsm8k	50861a	accuracy	gen	95.75	V1
> gsm8k	50861a	accuracy	gen	95.6	V2

Thanks for providing precision result of gsm8k.

**Review by freyfwt — APPROVED** [review]

**Review by Tflowers-0129 — APPROVED** [review]
